libavcodec/hevcdec.c

```diff
@@ -2654,6 +2654,221 @@ static int hls_slice_data_wpp(HEVCContext *s, const H2645NAL *nal)
+            s->HEVClcList[i] = av_mallocz(sizeof(HEVCLocalContext));
+            s->sList[i]->HEVClc = s->HEVClcList[i];
+        }
+    }
+
+    offset = (lc->gb.index >> 3);
+
+    for (j = 0, cmpt = 0, startheader = offset + s->sh.entry_point_offset[0]; j < nal->skipped_bytes; j++) {
+        if (nal->skipped_bytes_pos[j] >= offset && nal->skipped_bytes_pos[j] < startheader) {
+            startheader--;
+            cmpt++;
+        }
+    }
+    for (i = 1; i < s->sh.num_entry_point_offsets; i++) {
+        offset += (s->sh.entry_point_offset[i - 1] - cmpt);
+        for (j = 0, cmpt = 0, startheader = offset + s->sh.entry_point_offset[i]; j < nal->skipped_bytes; j++) {
+            if (nal->skipped_bytes_pos[j] >= offset && nal->skipped_bytes_pos[j] < startheader) {
+                startheader--;
+                cmpt++;
+            }
+        }
+        s->sh.size[i - 1] = s->sh.entry_point_offset[i] - cmpt;
+        s->sh.offset[i - 1] = offset;
+
+    }
+    if (s->sh.num_entry_point_offsets != 0) {
+        offset += s->sh.entry_point_offset[s->sh.num_entry_point_offsets - 1] - cmpt;
+        if (length < offset) {
+            av_log(s->avctx, AV_LOG_ERROR, "entry_point_offset table is corrupted\n");
+            res = AVERROR_INVALIDDATA;
+            goto error;
+        }
+        s->sh.size[s->sh.num_entry_point_offsets - 1] = length - offset;
+        s->sh.offset[s->sh.num_entry_point_offsets - 1] = offset;
+
+    }
+    s->data = data;
+
+    for (i = 1; i < s->threads_number; i++) {
+        /* Initialize the necessary parts of the local context object */
+        s->sList[i]->HEVClc->first_qp_group = 1;
+        s->sList[i]->HEVClc->qp_y = s->sList[0]->HEVClc->qp_y;
+        s->sList[i]->HEVClc->ctb_left_flag = s->sList[0]->HEVClc->ctb_left_flag ;
+        s->sList[i]->HEVClc->ctb_up_flag = s->sList[0]->HEVClc->ctb_up_flag ;
+        s->sList[i]->HEVClc->ctb_up_right_flag = s->sList[0]->HEVClc->ctb_up_right_flag ;
+        s->sList[i]->HEVClc->ctb_up_left_flag = s->sList[0]->HEVClc->ctb_up_left_flag ;
+        s->sList[i]->HEVClc->end_of_tiles_x = s->sList[0]->HEVClc->end_of_tiles_x ;
+        s->sList[i]->HEVClc->end_of_tiles_y = s->sList[0]->HEVClc->end_of_tiles_y ;
+        s->sList[i]->HEVClc->gb = s->sList[0]->HEVClc->gb;
+        memcpy(s->sList[i], s, sizeof(HEVCContext));
+        s->sList[i]->HEVClc = s->HEVClcList[i];
+    }
+
+    for (i = 0; i <= s->sh.num_entry_point_offsets; i++) {
+        arg[i] = i;
+        ret[i] = 0;
+    }
+
+    /* Parse the tiles in parallel */
+    if (s->ps.pps->tiles_enabled_flag) {
+        s->avctx->execute2(s->avctx, hls_decode_entry_tiles, arg, ret, s->sh.num_entry_point_offsets + 1);
+    }
+
+    for (i = 0; i <= s->sh.num_entry_point_offsets; i++) {
+        if (ret[i] < 0) {
+                res = ret[i];
+                goto error;
+        }
+        if (ret[i] > res) {
+            res = ret[i];
+        }
+    }
+
```

Comment:
we have to do the filtering at the end because it crosses tile boundaries, so both sides of a tile need to be done before the filtering.